src/graphnet/models/components/layers.py

```diff
@@ -593,3 +602,393 @@ def forward(
+            activation: Activation function.
+            norm_edges: Apply normalization to edges.
+            update_edges: Update edges after layer.
+            batch_norm_momentum: Momentum of batch normalization.
+            batch_norm_runner: Track running stats of batch normalization.
+            rezero: Apply learnable scaling parameters.
+            enable_edge_transform: Apply a FC to edges at the start
+                of the layer.
+            attn_bias: Add bias to keys and values in MHA block.
+            attn_dropout: Attention droput.
+            attn_clamp: Clamp absolute value of attention scores to a value.
+            attn_activation: Activation function for MHA block.
+            attn_edge_enhance: Applies learnable weight matrix with node-pair
+                in output node calculation in MHA block.
+        """
+        super().__init__()
+
+        self.in_channels = in_dim
+        self.out_channels = out_dim
+        self.in_dim = in_dim
+        self.out_dim = out_dim
+        self.num_heads = num_heads
+        self.dropout = dropout
+        self.residual = residual
+        self.update_edges = update_edges
+        self.batch_norm_momentum = batch_norm_momentum
+        self.batch_norm_runner = batch_norm_runner
+        self.rezero = rezero
+        self.deg_scaler = deg_scaler
+        self.activation = activation()
+
+        self.attention = GritSparseMHA(
+            in_dim=in_dim,
+            out_dim=out_dim // num_heads,
+            num_heads=num_heads,
+            use_bias=attn_bias,
+            dropout=attn_dropout,
+            clamp=attn_clamp,
+            activation=attn_activation,
+            edge_enhance=attn_edge_enhance,
+        )
+
+        self.fc1_x = nn.Linear(out_dim // num_heads * num_heads, out_dim)
+        if enable_edge_transform:
+            self.fc1_e = nn.Linear(out_dim // num_heads * num_heads, out_dim)
+        else:
+            self.fc1_e = nn.Identity()
+
+        if self.deg_scaler:
+            self.deg_coef = nn.Parameter(
+                torch.zeros(1, out_dim // num_heads * num_heads, 2)
+            )
+            nn.init.xavier_normal_(self.deg_coef)
+
+        if norm == nn.LayerNorm:
+            self.norm1_x = norm(out_dim)
+            self.norm1_e = self.norm(out_dim) if norm_edges else nn.Identity()
+        elif norm == nn.BatchNorm1d:
+            self.norm1_x = norm(
+                out_dim,
+                track_running_stats=self.batch_norm_runner,
+                eps=1e-5,
+                momentum=self.batch_norm_momentum,
+            )
+            self.norm1_e = (
+                norm(
+                    out_dim,
+                    track_running_stats=self.batch_norm_runner,
+                    eps=1e-5,
+                    momentum=self.batch_norm_momentum,
+                )
+                if norm_edges
+                else nn.Identity()
+            )
+        else:  # TODO: Maybe just set this to nn.Identity. -PW
```

Comment:
I think preferred to raise the error if the user passes a non-compatible layer instead of passing warnings and defaulting to identity